### PR TITLE
Updated Pest branch

### PR DESCRIPTION
**Pull request by dw2360**
I updated the pest branch to accept the # character and a multi-line string for the rule for a code block.
Approach: To differentiate a code block from an inline comment and a doc block, I have created a separate multi-line string variable that accepts any character within either the single triple quoted string or the double triple quoted string. The code doesn't recognize the comment and then a quote test case, but it does recognize the quote and then a comment test case.

### CODE REVIEW

**Comment by dw2360** [comment]
Currently working on the quote to comment test case

File: server/src/lexer/pest/python.pest [hunk 1]
```diff
@@ -82,7 +82,10 @@ block_comment_ending = { newline_eoi }
 
 // ## Code blocks
 code_block = { code_line+ }
-code_line = _{ (!doc_block ~ logical_line ~ NEWLINE) | (!doc_block ~ not_newline+ ~ EOI) }
+single_line_string = {("\"" ~ (!("\"") ~ ANY)* ~ "\"") | ("'" ~ (!("'") ~ ANY)* ~ "'")}
```
Comment (bjones1) [inline server/src/lexer/pest/python.pest:85]: This is a good start! First, get this working, then add some tricky extra cases:

``` Python
print("this is\
# not a comment.")
print("this is \" quoted") 
```

Output from running this:
```
this is# not a comment.
this is " quoted
```

File: server/src/lexer/pest/python.pest [hunk 2]
```diff
@@ -82,7 +82,10 @@ block_comment_ending = { newline_eoi }
...
+// If there is text or characters within the triple quoted string, it will be regarded as a code block.
+multi_line_string = { ("'''" ~ (!("'''") ~ ANY)* ~ "'''") | "\"\"\"" ~ (!("\"\"\"") ~ ANY)* ~ "\"\"\""}
+code_line = _{ (!single_line_string ~ NEWLINE | not_newline) | (multi_line_string ~ NEWLINE) | !(inline_comment_delim_0) ~ ("'''" | "\"\"\"") | (!doc_block ~ not_newline+ ~ EOI) }
```
Comment (bjones1) [inline server/src/lexer/pest/python.pest:88]: Not quite right. I think you want to say that code: not a comment, but is a single-line or multi-line string. Here, you say it's not a single/multi-line string.

File: server/src/lexer/pest/python.pest [hunk 3]
```diff
@@ -54,7 +54,7 @@ inline_comment_contents = { logical_line ~ newline_eoi }
 // [C standard, section 5.1.1.2](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3220.pdf#page=24),
 // if a line of code ends with a backslash, it continues on the next line. This
 // is a logical line; treat it as a single line.
-logical_line = _{ ("\\" ~ NEWLINE | not_newline)* }
+logical_line = _{ ("#" ~ NEWLINE | not_newline)* }
```
[the remark this answers is not in the document]
Comment (dw2360) [reply server/src/lexer/pest/python.pest]: Removed the logical line in new code

File: server/src/lexer/pest/python.pest (hunk 2 above)
Comment (dw2360) [reply server/src/lexer/pest/python.pest:90]: Yes, I was trying out some test cases and forgot to update the code since the single-line string was the only one recognized.

File: server/src/lexer/pest/python.pest [hunk 4]
```diff
@@ -20,12 +20,13 @@ file      =  { SOI ~ (doc_block | code_block)* }
...
 // Per the
-// [C standard, section 6.4.3](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3220.pdf#page=65),
-// "white-space consists of: (space, horizontal tab, new-line, vertical tab, and
+// [Python standard, section string.whitespace](https://docs.python.org/3/library/string.html#string.whitespace),
+// "white-space consists of: (space, horizontal tab, linefeed, return, vertical tab, and
 // form-feed)." Omit newlines, since the rest of this parser uses these.
 vertical_tab = { "\x0B" }
 form_feed    = { "\x0C" }
-white_space  = { (" " | "\t" | vertical_tab | form_feed)* }
+return       = {"\0x0D"}
```
Comment (bjones1) [inline server/src/lexer/pest/python.pest:28]: Omit -- this is covered by `NEWLINE` elsewhere in the grammar.

File: server/src/lexer/pest/python.pest [hunk 5]
```diff
@@ -20,12 +20,13 @@ file      =  { SOI ~ (doc_block | code_block)* }
...
+white_space  = { (" " | "\t" | vertical_tab | form_feed | return)* }
 
 // The
 // [C standard, section 6.4.9](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3220.pdf#page=65),
```
Comment (bjones1) [inline server/src/lexer/pest/python.pest:32]: Replace with a reference to the Python standard that discusses comments.

File: server/src/lexer/pest/python.pest [hunk 6]
```diff
@@ -43,34 +44,35 @@ inline_comment_cont = _{ PEEK[-1..-2] ~ PEEK ~ inline_comment_body }
...
+// A line ending in a backslash cannot carry a comment.
+// A backslash does not continue a comment. 
+// A backslash does not continue a token except for string literals
 newline_eoi  = _{ NEWLINE | EOI }
 not_newline  = _{ !NEWLINE ~ ANY }
 
 // ## Block comments
 //
 // Block comments never combine.
 block_comment = {
```
Comment (bjones1) [inline server/src/lexer/pest/python.pest:66]: Use this instead:

```
block_comment = { "unused" }
// Support multiple opening delimiters for other languages; assume the closing
// delimiter is always the same.
block_comment_opening_delim_0 =  { "unused" }
block_comment_opening_delim_1 =  { "unused" }
block_comment_opening_delim_2 =  { "unused" }
```

File: server/src/lexer/pest/python.pest [hunk 7]
```diff
@@ -82,7 +84,10 @@ block_comment_ending = { newline_eoi }
 
 // ## Code blocks
 code_block = { code_line+ }
-code_line = _{ (!doc_block ~ logical_line ~ NEWLINE) | (!doc_block ~ not_newline+ ~ EOI) }
+single_line_string = {("\"" ~ (!("\"") ~ ANY)* ~ "\"") | ("'" ~ (!("'") ~ ANY)* ~ "'")}
+// If there is text or characters within the triple quoted string, it will be regarded as a code block.
```
Comment (bjones1) [inline server/src/lexer/pest/python.pest:88]: Hints -- a code block is:
- Not a doc block, followed by non-newline characters then a newline OR
- Not a doc block, followed by non-newline characters  then EOI OR
- non-newline characters followed by a newline OR
- a multi-line string

**Review by bjones1 — CHANGES_REQUESTED** [review]

**Comment by bjones1** [comment]
Merge in commit 6f2e722b7470cf8c1696582ee82ac186ec71aaf4 to fix the tests.

File: server/src/lexer/pest/python.pest [hunk 8]
```diff
@@ -49,7 +49,7 @@ inline_comment_delim_2 =  { "unused" }
 // This accepts an inline comment containing a space then some contents, or an
 // inline comment immediately followed by a newline.
 inline_comment_body     = { (" " ~ inline_comment_contents) | newline_eoi }
-inline_comment_contents = { "#" ~ newline_eoi }
+inline_comment_contents = { ("#" ~ NEWLINE | not_newline)* ~ newline_eoi }
```
Comment (bjones1) [inline server/src/lexer/pest/python.pest:52]: Why this change? I seems redundant or wrong, I'm not sure which.

File: server/src/lexer/pest/python.pest [hunk 9]
```diff
@@ -59,20 +59,33 @@ inline_comment_contents = { "#" ~ newline_eoi }
...
+// Use this so that the ending (usually a newline) gets captured in a token.
+block_comment_ending = { newline_eoi }
 
 // ## Code blocks
 code_block = { code_line+ }
-single_line_string = {("\"" ~ (!("\"") ~ ANY)* ~ "\"") | ("'" ~ (!("'") ~ ANY)* ~ "'")}
 // If there is text or characters within the triple quoted string, it will be regarded as a code block.
 multi_line_string = { ("'''" ~ (!("'''") ~ ANY)* ~ "'''") | "\"\"\"" ~ (!("\"\"\"") ~ ANY)* ~ "\"\"\""}
-code_line = _{ (multi_line_string ~ NEWLINE | not_newline) | (single_line_string ~ NEWLINE) | !(inline_comment_delim_0) ~ ("'''" | "\"\"\"") | (!doc_block ~ not_newline+ ~ EOI) }
+code_line = _{ (!doc_block ~ not_newline+ ~ NEWLINE) | (multi_line_string ~ NEWLINE) | !(inline_comment_delim_0) ~ ("'''" | "\"\"\"") | (!doc_block ~ not_newline+ ~ EOI) }
```
Comment (bjones1) [inline server/src/lexer/pest/python.pest:88]: Still needs work -- have you talked with William?

**Review by bjones1 — CHANGES_REQUESTED** [review]
I need a working PR before we meet tomorrow. Feel free to schedule a meeting if you need more feedback.
